### ISSUE TITLE
unreliable raise-or-lower behavior with multiple workspaces

### ISSUE THREAD

**Issue by jmarino** [linuxmint/muffin#277]
I have mapped my windows key to "raise-or-lower" to easily switch between windows. I have done it using dconf-editor since this keybinding is not exposed in settings (why not?).

I found its behavior to be very unreliable. Sometimes it works fine: 
- press windows key on an obscured window and it is raised
- press windows key on an un-obscured window and it is lowered

However, sometimes it would seem to get stuck. I would press window key on an un-obscured window and nothing would happen.

I started digging into the code, function "handle_raise_or_lower (src/core/keybindings.c)", and found out that the stuck behavior happens when our window is compared against other windows in different workspaces. Thus, our window, which is on top in its own workspace, is not on top when compared with windows from other workspaces, and the code decides to raise it. Since it was already on top, nothing happens.

The source of the problem is that the stack of windows is traveled with "meta_stack_get_above", which seems to include windows from all workspaces. I admit I don't have a solid understanding of the code, so I don't know if this is supposed to be so, or if this is a sign that windows from different workspaces are being wrongly included in "window->screen->stack".

In any case, I include a patch that fixes the behavior of "raise-or-lower".

[fix-raise-or-lower.patch.gz](https://github.com/linuxmint/muffin/files/748399/fix-raise-or-lower.patch.gz)

**Comment by jmarino** [comment]
After some more testing I've noticed that the stack of windows that is traveled with "meta_stack_get_above" also includes minimized windows. The code is iterating through windows on different workspaces and also minimized windows. The patch in the previous comment only addresses the first issue.
Also, the previous patch doesn't do the right thing for windows that are visible in all workspaces.

A new version of the patch uses function "meta_window_should_be_showing" to decide if a window should be considered:

```
diff --git a/src/core/keybindings.c b/src/core/keybindings.c
index dcf295e..eb714ec 100644
--- a/src/core/keybindings.c
+++ b/src/core/keybindings.c
@@ -2963,7 +2963,7 @@ handle_raise_or_lower (MetaDisplay    *display,
     {
       MetaRectangle tmp, win_rect, above_rect;
       
-      if (above->mapped)
+      if (above->mapped && meta_window_should_be_showing(above))
         {
           meta_window_get_outer_rect (window, &win_rect);
           meta_window_get_outer_rect (above, &above_rect);
-- 
2.11.0
```

**Comment by jmarino** [comment]
Mutter is also affected  by this problem. I submitted the fix there too: bug [#705200](https://bugzilla.gnome.org/show_bug.cgi?id=705200).

**Comment by jmarino** [comment]
Fix merged with commit https://github.com/linuxmint/muffin/commit/130d6f00d4fb7ba468d462bf1447e8c3bc6d9dd8.
Closing issue.
